File: app/src/main/java/com/mohsenoid/rickandmorty/util/executor/IoTaskExecutor.java
package com.mohsenoid.rickandmorty.util.executor;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class IoTaskExecutor implements TaskExecutor {

    private static IoTaskExecutor instance;

    private final int THREADS_COUNT = 5;
    private final Executor executor = Executors.newFixedThreadPool(THREADS_COUNT);

    private IoTaskExecutor() { /* this will prevent making a new object of this type from outside */ }

    public static synchronized IoTaskExecutor getInstance() {
        if (instance == null)
            instance = new IoTaskExecutor();

        return instance;
    }

    @Override
    public void execute(Runnable runnable) {
        executor.execute(runnable);
    }
}

File: app/src/main/kotlin/com/mohsenoid/rickandmorty/util/executor/IoTaskExecutor.kt
package com.mohsenoid.rickandmorty.util.executor

import java.util.concurrent.Executor
import java.util.concurrent.Executors

class IoTaskExecutor : TaskExecutor {

    private val executor: Executor = Executors.newFixedThreadPool(THREADS_COUNT)

    override fun execute(runnable: () -> Unit) {
        executor.execute(runnable)
    }

    companion object {
        private const val THREADS_COUNT = 5
    }
}
